jax/interpreters/xla.py

```diff
@@ -149,23 +149,39 @@ def _make_abstract_python_scalar(typ, _):
 
 def apply_primitive(prim, *args, **params):
   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
-  abstract_args = map(abstractify, args)
-  compiled_fun = xla_primitive_callable(prim, *abstract_args, **params)
+  compiled_fun = xla_primitive_callable(prim, *map(arg_spec, args), **params)
   return compiled_fun(*args)
 
+def arg_spec(x):
+  aval = abstractify(x)
+  try:
+    return aval, x._device
+  except:
+    return aval, None
+
 @cache()
-def xla_primitive_callable(prim, *abstract_args, **params):
+def xla_primitive_callable(prim, *arg_specs, **params):
   backend = params.get('backend', None)
-  aval_out = prim.abstract_eval(*abstract_args, **params)
+  avals, devices = unzip2(arg_specs)
+  # TODO(mattjj): make Device hashable instead of handling pairs here
+  try:
+    device, = set(d for d in devices if d is not None) or (None,)
```

Comment:
Since this is on the eager dispatch hot path it might eventually make sense to have something like `get_common_device` in the XLA client?

Reply:
I don't think it's exactly on the hot path; it's under the cache. Moreover, it takes about ~316ns for binary inputs and 305ns for unary inputs, so I don't think it'll be a bottleneck in the near future (compilation time, like we do under the cache here, is likely to swamp it!).

Reply:
Oh yeah, makes sense. (It might make cache lookup take longer, though--and we don't _actually_ need to recompile for new devices, only for new platforms)

Reply:
I'm operating under the constraints of the current XLA client. Maybe there are future improvements to make there, though they are unaffected by this change.

The highlighted line here won't make cache lookup slower because it's underneath the memoized function. This PR overall makes cache lookup slower, but only by about 50ns.